### PR TITLE
Issue#63183: Add fs::read_dir() and ReadDir warning about iterator order + example

### PR DESCRIPTION
**Pull request by ali-raheem**
As per https://github.com/rust-lang/rust/issues/63183

Add warning about iterator order to read_dir and ReadDir, add example of explicitly ordering direntrys.

### CODE REVIEW

File: src/libstd/fs.rs [hunk 1]
```diff
@@ -114,6 +114,11 @@ pub struct Metadata(fs_imp::FileAttr);
 /// information like the entry's path and possibly other metadata can be
 /// learned.
 ///
+/// #### Note: Iteration Order is Implementation-Defined
```
Comment (Centril) [inline src/libstd/fs.rs]: I would remove this; it feels redundant with the more specific sentence below. Moreover, "implementation defined" means that each implementation must define the semantics and this feels more "unspecified".

Comment (ali-raheem) [reply src/libstd/fs.rs]: I guess if we are removing one (line 117) makes sense to also remove the other.

Comment (Centril) [reply src/libstd/fs.rs]: haha, that was implied ;)

File: src/libstd/fs.rs [hunk 2]
```diff
@@ -1991,6 +1997,28 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
 ///     Ok(())
 /// }
 /// ```
+///
+/// ```rust,no_run
+/// use std::{fs, io};
+///
+/// fn main() -> io::Result<()> {
+///     // The order read_dir returns entries is not guaranteed. If reproducible
```
Comment (abonander) [inline src/libstd/fs.rs]: Minor grammatical error here. Maybe you meant:

> The order _in which_ \`read_dir\` returns entries is not guaranteed.

Comment (abonander) [reply src/libstd/fs.rs]: Backticks aren't processed in comments but IMO it still helps visually demarcate that you're talking about a Rust item. I'm not sure if that fits the rust-lang code style. @Centril?

Comment (Centril) [reply src/libstd/fs.rs]: Yeah we prefer backticks when referring to code and code entities.

**Comment by gagan0723** [comment]
Ping from triage, @KodrAus can you please review this.

Thanks.

**Comment by KodrAus** [comment]
Thanks for your patience @ali-raheem! I think this example could be closer to a typical usecase if we retain the `DirEntry` itself, and sorting by the path:

```rust
let mut entries = fs::read_dir(".")?.collect::<Result<Vec<_>, io::Error>>()?;
        
entries.sort_by_key(|e| e.path());
```

That way the `entries` collection contains the whole `DirEntry` along with easy access to other filesystem metadata. It means that we couldn't use the `println!` statements so easily, but I don't think we really need them.

**Comment by abonander** [comment]
@KodrAus `DirEntry::path()` performs a copy, though.

**Comment by KodrAus** [comment]
Ah yes that's not really ideal is it. Thanks @abonander!

Ok I think this will be good to go once we remove the `println`s (or convert them into comments).

**Comment by ali-raheem** [comment]
@KodrAus Like this? I've commented out the printlns so if someone wanted they could uncomment them and run it as a demonstration.

**Comment by ProgrammaticNajel** [comment]
Ping from triage. @KodrAus any updates on this? Thanks.

**Comment by KodrAus** [comment]
Ah sorry @ali-raheem, I meant that I think we should just remove those statements altogether, but could have a comment in the example source like:

```rust
fn main() -> io::Result<()> {
    let mut entries = fs::read_dir(".")?
        .map(|res| res.map(|e| e.path()))
        .collect::<Result<Vec<_>, io::Error>>()?;

    // The order in which `read_dir` returns entries is not guaranteed. If reproducible
    // ordering is required the entries should be explicitly sorted.

    entries.sort();

    // The contents of `entries` are now sorted by their path

    Ok(())
}
```

**Comment by abonander** [comment]
The entire reason for opening the issue is that there should be a clear statement that the user should not rely on iteration order; comments in examples are too easy to ignore.

**Comment by KodrAus** [comment]
@abonander Hm, I would think prose within the example would be clearer for someone scanning it than `println` statements they can’t actually see the result of.

**Comment by abonander** [comment]
Ah, my mistake. I thought you were referring to the statements in the actual doc comments and I just immediately thought "No! That was my whole complaint to begin with!"

**Comment by ali-raheem** [comment]
@KodrAus 

Sorry it occured to me you might not get a notification when I push a new commit so I thought I'd ping. 

Hopefully made the changes now.

**Comment by ProgrammaticNajel** [comment]
Ping from triage. @KodrAus any updates on this? Thanks. 😄

**Review by KodrAus — APPROVED** [review]
Thanks for your patience @ali-raheem!

This looks good to me

**Pull request merged**